fdbserver/KeyValueStoreMemory.actor.cpp

```diff
@@ -485,19 +487,35 @@ class KeyValueStoreMemory : public IKeyValueStore, NonCopyable {
 						StringRef p1 = data.substr(0, h.len1);
 						StringRef p2 = data.substr(h.len1, h.len2);
 
-						if (h.op == OpSnapshotItem) { // snapshot data item
+						if (h.op == OpSnapshotItem || h.op == OpSnapshotItemDelta) { // snapshot data item
 							/*if (p1 < uncommittedNextKey) {
 								TraceEvent(SevError, "RecSnapshotBack", self->id)
 									.detail("NextKey", uncommittedNextKey)
 									.detail("P1", p1)
 									.detail("Nextlocation", self->log->getNextReadLocation());
 							}
 							ASSERT( p1 >= uncommittedNextKey );*/
+							if(h.op == OpSnapshotItemDelta) {
+								ASSERT(p1.size() > 1);
+								// Get number of bytes borrowed from previous item key
+								int borrowed = *(uint8_t *)p1.begin();
+								ASSERT(borrowed <= lastSnapshotKey.size());
+								// Trim p1 to just the suffix
+								StringRef suffix = p1.substr(1);
+								// Allocate a new string in data arena to hold prefix + suffix
+								Arena &dataArena = *(Arena *)&data.arena();
```

Comment:
Please use `const_cast<>()` to make why you're doing this more explicit.  I'm not actually sure if you can do:

```suggestion
								Arena &dataArena = const_cast<Arena&>(data.arena());
```

or if you do need to keep going to pointer and back.

Reply:
Actually, I'll merge, and fold this into my follow-on PR.

Reply:
We could also fix this more generally to not require the cast by changing `Standalone` to declare its `Arena` as `mutable` and return it as a non-const reference from `arena()` even for a `const Standalone<T>`.  I think this is defensible because conceptually the `T` is the `const` thing we aren't changing, the `Arena` just happens to be where it is located.  The idea of a `const Arena` would still exist for contexts where an Arena should not be added to for some reason, `const Standalone<T>` just would not use it.

I can't find any place in the code where this matters though.

Oh and I just realized another way to do it would be at the top of this small `if` block for Item opcodes just do `Arena dataArena(data.arena());` and then use `dataArena` instead of `data.arena()` in all places in the block.  This should compile to something that executes identically to the const_cast because the construction is a trivial copy of the argument, and Arena's sole member is a pointer.